pkg/build/util/util_test.go

```diff
@@ -287,22 +304,99 @@ func TestSafeForLoggingBuild(t *testing.T) {
 	if credsRegex.MatchString(*stripped.Spec.Source.Git.HTTPSProxy) {
 		t.Errorf("credentials left in https proxy value: %v", stripped.Spec.Source.Git.HTTPSProxy)
 	}
-	checkEnv(t, stripped.Spec.Strategy.SourceStrategy.Env)
-	checkEnv(t, stripped.Spec.Strategy.DockerStrategy.Env)
-	checkEnv(t, stripped.Spec.Strategy.CustomStrategy.Env)
-	checkEnv(t, stripped.Spec.Strategy.JenkinsPipelineStrategy.Env)
+	checkEnv(t, stripped.Spec.Strategy.SourceStrategy.Env, false)
+	checkEnv(t, stripped.Spec.Strategy.DockerStrategy.Env, false)
+	checkEnv(t, stripped.Spec.Strategy.CustomStrategy.Env, false)
+	checkEnv(t, stripped.Spec.Strategy.JenkinsPipelineStrategy.Env, false)
+
+	// make sure original object was not touched
+	if redactedRegex.MatchString(*proxyBuild.Spec.Source.Git.HTTPProxy) {
+		t.Errorf("improperly redacted credentials in original http proxy value: %s", *stripped.Spec.Source.Git.HTTPProxy)
+	}
+	if redactedRegex.MatchString(*proxyBuild.Spec.Source.Git.HTTPSProxy) {
+		t.Errorf("improperly redacted credentials in https proxy value: %s", *stripped.Spec.Source.Git.HTTPSProxy)
+	}
+	checkEnv(t, proxyBuild.Spec.Strategy.SourceStrategy.Env, true)
+	checkEnv(t, proxyBuild.Spec.Strategy.DockerStrategy.Env, true)
+	checkEnv(t, proxyBuild.Spec.Strategy.CustomStrategy.Env, true)
+	checkEnv(t, proxyBuild.Spec.Strategy.JenkinsPipelineStrategy.Env, true)
 }
 
-func checkEnv(t *testing.T, envs []corev1.EnvVar) {
+func checkEnv(t *testing.T, envs []corev1.EnvVar, orig bool) {
 	for _, env := range envs {
 		if env.Name == "other_value" {
 			if !credsRegex.MatchString(env.Value) {
 				t.Errorf("credentials improperly stripped from env value %v", env)
 			}
 		} else {
-			if credsRegex.MatchString(env.Value) {
-				t.Errorf("credentials not stripped from env value %v", env)
+			if orig {
+				if redactedRegex.MatchString(env.Value) {
+					t.Errorf("credentials improperly stripped from original env value %v", env)
+				}
+				if isURL(env.Value) && urlStringHasPassword(env.Value) && !credsRegex.MatchString(env.Value) {
+					t.Errorf("credentials missing from original env value %v", env)
+				}
+			} else {
+				if credsRegex.MatchString(env.Value) {
+					t.Errorf("credentials not stripped from env value %v", env)
+				}
+				if isURL(env.Value) && urlStringHasPassword(env.Value) && !redactedRegex.MatchString(env.Value) {
+					t.Errorf("redacted should appear in env value %v", env)
+				}
 			}
 		}
 	}
 }
+
+func isURL(rawurl string) bool {
+	_, err := url.Parse(rawurl)
+	return err != nil
+}
+
+func urlStringHasPassword(rawurl string) bool {
+	url, err := url.Parse(rawurl)
+	if err != nil {
+		return false
+	}
+	return urlHasPassword(url)
+}
+
+func urlHasPassword(url *url.URL) bool {
+	if url.User == nil {
+		return false
+	}
+	_, hasPassword := url.User.Password()
+	return hasPassword
+}
+
+func TestSafeForLoggingURL(t *testing.T) {
+	cases := []string{
+		"https://user:password@hostname.com",
+		"https://www.redhat.com",
```

Comment:
Explicit unit test for `SafeForLoggingURL` - other cases to add?

Reply:
user w/ no password?

Reply:
(sorry if these were here before and i missed them)